app/services/content_graph_api.py

```
@@ -0,0 +1,161 @@
+"""
+Interactions with the Content Graph API, including transformation
+to Snowdrop format.
+"""
+import logging
+import os
+from enum import Enum
+from urllib.parse import urljoin
+from urllib.request import pathname2url
+
+import requests
+
+
+DEFAULT_HOST = "http://content-graph.stage.api.bbci.co.uk"
+HOST = os.getenv("CONTENT_GRAPH_API_HOST", DEFAULT_HOST)
+
+ITEM_URL = urljoin(HOST, 'content/{}')
+SIMILAR_URL = urljoin(HOST, 'content/{}/similar?limit={}&mediaType={}')
+RANDOM_URL = urljoin(HOST, 'content?random=true&limit={}&mediaType={}')
+
+session = requests.Session()
+logger = logging.getLogger("snowdrop")
+logger.info('CONTENT_GRAPH_API_HOST: {}'.format(HOST))
+
+
+class MediaType(Enum):
+    VIDEO = "video"
+
+    def __str__(self):
+        return self.value
+
+
+class ContentGraphAPIException(Exception):
+    """Exceptions raised while interacting with the Content Graph API."""
+    pass
+
+
+def pid_to_uri(pid):
+    """Return the URI given the PID of a Video Program."""
+    return "programmes:bbc.co.uk,2018/FIXME/{}".format(pid)
+
+
+def to_snowdrop(cg_item, version=True):
+    """
+    Convert a content graph media item dict into a Snowdrop
+    media item dict.
+
+    Sample output:
+    {
+        "image": {
+            "altText": "Sample image",
+            "urlTemplate": "http://ichef...",
+        },
+            "id": "p123gas1v",
+            "title": "Sample video",
+            "versionId": "p1jf934"
+    }
+    """
+    try:
+        sd_item = {
+            "image": {
+                "altText": cg_item["Title"],
+                "urlTemplate": cg_item["Image"],
+            },
+            "id": cg_item["Pid"],
+            "title": cg_item["Title"]
+        }
+        if version:
+            sd_item["versionId"] = cg_item["Version"]
+    except KeyError:
+        msg = {
+            "error": "Unable to convert item from the CG API",
+            "details": {
+                "item": cg_item
```

Just wondering if we could echo the item back to the caller? Maybe it'd be worth, eventually, logging it but returning it might not be as safe?

---

In some projects people don't want to expose API error details to the clients, in others they want. I'm fine with both approaches :) I'll change to return this in the payload of the response.

---

More just in case there is an issue with the content of `cg_item` and it somehow exposes something. I'm thinking more along an OWASP sense, as: https://www.owasp.org/index.php/Error_Handling

I think we should return the error message but maybe not echoing the value back again. It would be useful to debug so should be (eventually) in the logs maybe with a unique ID etc.

---

Yes, @betandr , it would be great if we could have a unique ID in the user request - and we could propagate this throughout our services. It would make debugging so much easier!

---

As we talked, @betandr, I'll leave as it is for now - until we make a decision on how much we want to expose the errors to the clients.